File: 093/Sliding_Window_Maximum.c
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int maxTwo(int a, int b) {
    return a > b ? a : b;
}

int* maxSlidingWindow(int* nums, int numsSize, int k, int* returnSize) {
    if (!nums || numsSize == 0) return NULL;
    *returnSize = numsSize - k + 1;
    int *ret = (int *)malloc(*returnSize * sizeof(int));
    if (!ret) return NULL;

    int *maxFromLeft = (int *)malloc(numsSize * sizeof(int));
    if (!maxFromLeft) return NULL;
    int *maxFromRight = (int *)malloc(numsSize * sizeof(int));
    if (!maxFromRight) return NULL;

    maxFromLeft[0] = nums[0];
    maxFromRight[numsSize-1] = nums[numsSize-1];

    int i, j;
    for (i = 1, j = numsSize - 2; i < numsSize; i++, j--) {
        maxFromLeft[i] = i % k == 0 ? nums[i] : maxTwo(maxFromLeft[i-1], nums[i]);
        maxFromRight[j] = j % k == k - 1 ? nums[j] : maxTwo(maxFromRight[j+1], nums[j]);
    }
    for (i = 0; i < *returnSize; i++) {
        ret[i] = maxTwo(maxFromLeft[i+k-1], maxFromRight[i]);
    }
    free(maxFromLeft);
    free(maxFromRight);
    return ret;
}

File: 239/Sliding_Window_Maximum.go
func max(a int, b int) int {
    if a > b {
        return a
    } else {
        return b
    }
}

func maxSlidingWindow(nums []int, k int) []int {
    if k == 0 || nums == nil {
        return nil
    }
    length := len(nums)

    var ret []int

    maxFromLeft := make([]int, length)
    maxFromRight := make([]int, length)

    maxFromLeft[0] = nums[0]
    maxFromRight[length - 1] = nums[length - 1]

    for i, j := 1, length - 2; i < length; i, j = i + 1, j - 1 {
        if i % k == 0 {
            maxFromLeft[i] = nums[i]
        } else {
            maxFromLeft[i] = max(maxFromLeft[i - 1], nums[i])
        }

        if j % k == k - 1 {
            maxFromRight[j] = nums[j]
        } else {
            maxFromRight[j] = max(maxFromRight[j + 1], nums[j])
        }
    }

    for i := 0; i < length - k + 1; i++ {
        ret = append(ret, max(maxFromLeft[i + k - 1], maxFromRight[i]))
    }

    return ret
}
